# The DataFrame and Series

The DataFrame and Series are the two primary pandas objects that we will be using throughout this book.

* **DataFrame** - A two-dimensional data structure that looks like any other rectangular table of data you have seen with rows and columns.
* **Series** - A single dimension of data. It is analogous to a single column of data or a one dimensional array.

## Import pandas and read in data with `read_csv`

By convention, pandas is imported and aliased as `pd`. We will read in the `bikes` dataset with the `read_csv` function. Its first parameter is the location of the file relative to the current directory as a string. All of the data for this book is stored in the `data` directory one level above where this notebook is located. The two dots in the path passed to `read_csv` are interpreted as the directory immediately above the current one.

[1]: images/pandas_logo.png
[2]: http://stackoverflow.com/questions/tagged/pandas
[3]: https://www.youtube.com/watch?v=kHdkFyGCxiY
[4]: http://www.numpy.org/

In [1]:
import pandas as pd
bikes = pd.read_csv('../data/bikes.csv')

### Display DataFrame in Jupyter Notebook
We assigned the output from the `read_csv` function to the `bikes` variable which now refers to our DataFrame object. To visually display the DataFrame, place the variable name as the last line in a code cell. By default, pandas outputs 60 rows and 20 columns. 

### `head` and `tail` methods
A very useful and simple method is `head`, which returns the first 5 rows of the DataFrame by default. This avoids long default output and is something I highly recommend when doing data analysis within a notebook. The `tail` method returns the last 5 rows by default. There will be very few instances in the book where the `head` method is not used as displaying 60 rows is far too many and will take up a lot of space on a screen or page.

In [2]:
bikes.head()

trip_id    usertype gender            starttime             stoptime  \
0     7147  Subscriber   Male  2013-06-28 19:01:00  2013-06-28 19:17:00   
1     7524  Subscriber   Male  2013-06-28 22:53:00  2013-06-28 23:03:00   
2    10927  Subscriber   Male  2013-06-30 14:43:00  2013-06-30 15:01:00   
3    12907  Subscriber   Male  2013-07-01 10:05:00  2013-07-01 10:16:00   
4    13168  Subscriber   Male  2013-07-01 11:16:00  2013-07-01 11:18:00   

   tripduration             from_station_name  latitude_start  \
0           993     Lake Shore Dr & Monroe St       41.881050   
1           623  Clinton St & Washington Blvd       41.883380   
2          1040  Sheffield Ave & Kingsbury St       41.909592   
3           667       Carpenter St & Huron St       41.894556   
4           130        Damen Ave & Pierce Ave       41.909396   

   longitude_start  dpcapacity_start          to_station_name  latitude_end  \
0       -87.616970              11.0    Michigan Ave & Oak St     41.900960   
1       -87.641170              31.0     Wells St & Walton St     41.899930   
2       -87.653497              15.0  Dearborn St & Monroe St     41.881320   
3       -87.653449              19.0   Clark St & Randolph St     41.884576   
4       -87.677692              19.0   Damen Ave & Pierce Ave     41.909396   

   longitude_end  dpcapacity_end  temperature  visibility  wind_speed  \
0     -87.623777            15.0         73.9        10.0        12.7   
1     -87.634430            19.0         69.1        10.0         6.9   
2     -87.629521            23.0         73.0        10.0        16.1   
3     -87.631890            31.0         72.0        10.0        16.1   
4     -87.677692            19.0         73.0        10.0        17.3   

   precipitation        events  
0        -9999.0  mostlycloudy  
1        -9999.0  partlycloudy  
2        -9999.0  mostlycloudy  
3        -9999.0  mostlycloudy  
4        -9999.0  partlycloudy

The last five rows of the DataFrame are displayed with the `tail` method.

In [3]:
bikes.tail()

trip_id    usertype  gender            starttime             stoptime  \
50084  17534938  Subscriber    Male  2017-12-30 13:07:00  2017-12-30 13:34:00   
50085  17534969  Subscriber    Male  2017-12-30 13:34:00  2017-12-30 13:44:00   
50086  17534972  Subscriber    Male  2017-12-30 13:34:00  2017-12-30 13:48:00   
50087  17535645  Subscriber  Female  2017-12-31 09:30:00  2017-12-31 09:33:00   
50088  17536246  Subscriber    Male  2017-12-31 15:22:00  2017-12-31 15:26:00   

       tripduration           from_station_name  latitude_start  \
50084          1625       State St & Pearson St       41.897448   
50085           585    Halsted St & 35th St (*)       41.830661   
50086           824    Kingsbury St & Kinzie St       41.889177   
50087           178        Clinton St & Lake St       41.885637   
50088           214  Clarendon Ave & Leland Ave       41.967968   

       longitude_start  dpcapacity_start                to_station_name  \
50084       -87.628722              27.0              Clark St & Elm St   
50085       -87.647172              16.0            Union Ave & Root St   
50086       -87.638506              31.0  Halsted St & Blackhawk St (*)   
50087       -87.641823              23.0       Kingsbury St & Kinzie St   
50088       -87.650001              15.0     Clifton Ave & Lawrence Ave   

       latitude_end  longitude_end  dpcapacity_end  temperature  visibility  \
50084     41.902973     -87.631280            27.0          5.0        10.0   
50085     41.819102     -87.643278            11.0          5.0        10.0   
50086     41.908537     -87.648627            20.0          5.0        10.0   
50087     41.889177     -87.638506            31.0          7.0        10.0   
50088     41.968812     -87.657659            15.0         10.9        10.0   

       wind_speed  precipitation        events  
50084        16.1        -9999.0  partlycloudy  
50085        16.1        -9999.0  partlycloudy  
50086        16.1        -9999.0  partlycloudy  
50087        11.5        -9999.0  partlycloudy  
50088        15.0        -9999.0  partlycloudy

### First and last `n` rows
Both the `head` and `tail` methods take a single integer parameter `n` controlling the number of rows returned. Here, we output the first three rows.

In [4]:
bikes.head(3)

trip_id    usertype gender            starttime             stoptime  \
0     7147  Subscriber   Male  2013-06-28 19:01:00  2013-06-28 19:17:00   
1     7524  Subscriber   Male  2013-06-28 22:53:00  2013-06-28 23:03:00   
2    10927  Subscriber   Male  2013-06-30 14:43:00  2013-06-30 15:01:00   

   tripduration             from_station_name  latitude_start  \
0           993     Lake Shore Dr & Monroe St       41.881050   
1           623  Clinton St & Washington Blvd       41.883380   
2          1040  Sheffield Ave & Kingsbury St       41.909592   

   longitude_start  dpcapacity_start          to_station_name  latitude_end  \
0       -87.616970              11.0    Michigan Ave & Oak St      41.90096   
1       -87.641170              31.0     Wells St & Walton St      41.89993   
2       -87.653497              15.0  Dearborn St & Monroe St      41.88132   

   longitude_end  dpcapacity_end  temperature  visibility  wind_speed  \
0     -87.623777            15.0         73.9        10.0        12.7   
1     -87.634430            19.0         69.1        10.0         6.9   
2     -87.629521            23.0         73.0        10.0        16.1   

   precipitation        events  
0        -9999.0  mostlycloudy  
1        -9999.0  partlycloudy  
2        -9999.0  mostlycloudy

## Components of a DataFrame

The DataFrame is composed of three separate components - the **columns**, the **index**, and the **data**. These terms will be used throughout the book and understanding them is vital to your ability to use pandas. Take a look at the following graphic of our `bikes` DataFrame stylized to put emphasis on each component.

![][1]

[1]: images/df_components.png

### The index
The index provides a **label** for each row and will always be displayed in **bold** font to the left of the data. For instance, the index label **3** references all the values in its row.

The index is also referred to as the **row labels** and individual values in the index are referred to as an **index label** or **row label**. A major purpose of the index is to identify each row in the DataFrame. Similarly to how we humans have names to identify ourselves, the index has labels to identify each row.

In the above DataFrame, the index is simply a sequence of integers beginning at 0. The values in the index are not limited to integers. Strings are a common type that are used in the index and make for more descriptive labels.

Surprisingly, values in the index are not required to be unique. In fact, all the index values can be the same. This is important to understand, as knowing a row label does not guarantee a one-to-one mapping to one specific row.

### The columns
The columns provide a **label** for each column and will always be displayed in **bold** font above the data. In the above DataFrame, the column name `from_station_name` references all the values in that column.

The columns are also referred to as the **column names** or the **column labels** with individual values referred to as a **column name** or **column label**.

The column names in the above DataFrame are strings, but can be other types such as integers though string names are more descriptive. Like the index, pandas will default column names to an integer sequence beginning at 0, if the columns are not provided upon DataFrame creation.

As with the index, the column names are not required to be unique, though having duplicate columns would be bad practice as its vital to be able to uniquely identify each column.

### The data
The actual data is to the right of the index and below the columns and is always in normal font. The data is also referred to as the **values**. The data represents all the values for all the columns. It is important to note that the index and the columns are NOT part of the data. They are separate objects and act as **labels**.

### The Axes
The index and columns are known collectively as the **axes** and each represent a single **axis** of the two-dimensional DataFrame. pandas uses the integer **0** for the index and <strong>1</strong> for the columns. pandas allows you to use these integers when referencing the index and columns within methods. This terminology is borrowed from numpy's ndarray objects which can have any number of dimensions.

## What type of object is `bikes`?
Let's verify that `bikes` is indeed a DataFrame with the `type` function.

In [5]:
type(bikes)

pandas.core.frame.DataFrame

### Fully-qualified name
The above output returns something called the **fully-qualified name**. Only the word after the last dot is the name of the type and in this case we have verified that the `bikes` variable has type `DataFrame`. The fully-qualified name always returns the package and module name of where the type was defined.

The package name is the first part of the fully-qualified name and in this case is `pandas`. The module name is the word immediately preceding the name of the type and in this case is `frame`.

### Package vs Module
A Python **package** is a directory containing other directories or modules. A Python **module** is a file (typically a text file ending in .py) that contains Python code of where the type is defined (typically with a `class` statement).

### Sub-packages
Any directory containing other directories or modules within a Python package is considered a **sub-package**. In this case, `core` is the sub-package.

### Where are the packages located on my machine?
Third-party packages are installed in the `site-packages` directory which itself is set up during Python installation. We can get the actual location with the help from the standard library's `site` module's `getsitepackages` function.

In [ ]:
import site
sp = site.getsitepackages()
sp

You can navigate to this directory with your file explorer. The will be a 'pandas' directory found there and within that will be a 'core' directory which will contain the 'frame.py' file. It is this text file which contains Python code where the DataFrame class is defined.

## Select a single column from a DataFrame - a Series
To select a single column from a DataFrame, pass the name of one of the columns to the brackets operator, `[]`. The returned object will be a pandas **Series**. Let's select the column name `tripduration`, assign it to a variable, and output the first few values to the screen. The `head` and `tail` methods work the same as they do with DataFrames.

In [7]:
trip_duration = bikes['tripduration']
trip_duration.head()

0     993
1     623
2    1040
3     667
4     130
Name: tripduration, dtype: int64

In [8]:
trip_duration.tail(3)

50086    824
50087    178
50088    214
Name: tripduration, dtype: int64

## Components of a Series - the index and the data
A Series is simpler than a DataFrame with just a single dimension of data. It has two components - the **index** and the **data**. Let's take a look at a stylized Series graphic.

![](images/series_components.png)

It's important to note that a Series has no rows and no columns. In appearance, it resembles a one-column DataFrame, but it technically has no columns. It just has a sequence of values that are labeled by an index.

### The index

A Series index serves as labels for the values and a single index **label** always references a single value. In the above image, the index label **3** corresponds with the value 667. The Series index is virtually identical to the DataFrame index, so the same rules apply to it. Index values can be duplicated and can be types other than integers, such as strings. 

### Output of Series vs DataFrame
Notice that there is no nice HTML styling for the Series. It's just plain text. Also, below each Series will be some metadata on it - the **name**, **length**, and **dtype**. 

* The **name** is not important right now. If the Series is formed from a column of a DataFrame it will be set to that column name.
* The **length** is the number of values in the Series
* The **dtype** is the data type of the Series, which will be discussed in upcoming chapters.

It's important to note that this metadata is NOT part of the Series itself and is just some extra info pandas outputs for your information.

## Exercises
Use the `bikes` DataFrame for the following exercises.

### Exercise 1
<span  style="color:green; font-size:16px">Select the column `events`, the type of weather that was recorded, and assign it to a variable with the same name. Output the first 10 values of it.</span>

In [9]:
bikes['events'].head(10)

0    mostlycloudy
1    partlycloudy
2    mostlycloudy
3    mostlycloudy
4    partlycloudy
5    mostlycloudy
6          cloudy
7          cloudy
8          cloudy
9    mostlycloudy
Name: events, dtype: object

### Exercise 2
<span  style="color:green; font-size:16px">What type of object is `events`?</span>

In [10]:
bikes.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50089 entries, 0 to 50088
Data columns (total 19 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   trip_id            50089 non-null  int64  
 1   usertype           50089 non-null  object 
 2   gender             50089 non-null  object 
 3   starttime          50089 non-null  object 
 4   stoptime           50089 non-null  object 
 5   tripduration       50089 non-null  int64  
 6   from_station_name  50089 non-null  object 
 7   latitude_start     50083 non-null  float64
 8   longitude_start    50083 non-null  float64
 9   dpcapacity_start   50083 non-null  float64
 10  to_station_name    50089 non-null  object 
 11  latitude_end       50077 non-null  float64
 12  longitude_end      50077 non-null  float64
 13  dpcapacity_end     50077 non-null  float64
 14  temperature        50089 non-null  float64
 15  visibility         50089 non-null  float64
 16  wind_speed         500

### Exercise 3
<span  style="color:green; font-size:16px">Select the last 2 rows of the `bikes` DataFrame and assign it to the variable `bikes_last_2`. What type of object is `bikes_last_2`?</span>

In [12]:
bikes_last_2 = bikes.iloc[:, :-2]
bikes_last_2.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50089 entries, 0 to 50088
Data columns (total 17 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   trip_id            50089 non-null  int64  
 1   usertype           50089 non-null  object 
 2   gender             50089 non-null  object 
 3   starttime          50089 non-null  object 
 4   stoptime           50089 non-null  object 
 5   tripduration       50089 non-null  int64  
 6   from_station_name  50089 non-null  object 
 7   latitude_start     50083 non-null  float64
 8   longitude_start    50083 non-null  float64
 9   dpcapacity_start   50083 non-null  float64
 10  to_station_name    50089 non-null  object 
 11  latitude_end       50077 non-null  float64
 12  longitude_end      50077 non-null  float64
 13  dpcapacity_end     50077 non-null  float64
 14  temperature        50089 non-null  float64
 15  visibility         50089 non-null  float64
 16  wind_speed         500